transaction/core/src/memo.rs

```diff
@@ -0,0 +1,153 @@
+//! Definition of memo payload type
+//!
+//! The encrypted memo of TxOut's is designed to have one encryption scheme and
+//! the payload is an extensible format. Two bytes are used for a schema type,
+//! and thirty two bytes are used for data according to that schema.
+//!
+//! The encryption details are defined in the transaction crate, but we would
+//! like to avoid making the introduction of a new schema require changes to
+//! the transaction-core crate, because this would require a new consensus
+//! enclave.
+//!
+//! We also would like to avoid implementing the interpretation of memo data
+//! in the transaction crate, for much the same reasons.
+//!
+//! Therefore, the code is organized as follows:
+//! - A MemoPayload is the collection of 34 bytes ready to be encrypted. This
+//!   can be used to construct a TxOut, and it is encrypted at that time. This
+//!   is defined in transaction-core crate.
+//! - The memo module in transaction-std crate defines specific structures that
+//!   can be converted to a MemoPayload, and provides a function that can
+//!   interpret a MemoPayload as one of the known high-level objects.
+//! - The TransactionBuilder has optional values that can be set on it to set
+//!   the "policy" around memos for this transaction, so that low-level handling
+//!   of memos is not needed by the user of the TransactionBuilder.
+//! - When interpretting memos on TxOut's that you recieved, the memo module
+//!   functionality can be used to assist.
+
+use aes::{
+    cipher::{FromBlockCipher, StreamCipher},
+    Aes256, Aes256Ctr, NewBlockCipher,
+};
+use core::convert::{TryFrom, TryInto};
+use displaydoc::Display;
+use generic_array::{
+    sequence::Split,
+    typenum::{U32, U34, U48},
+    GenericArray,
+};
+use hkdf::Hkdf;
+use mc_crypto_keys::{CompressedRistrettoPublic, RistrettoPublic};
+use serde::{Deserialize, Serialize};
+use sha2::Sha512;
+
+/// A plaintext memo payload, with accessors to easily access the memo type
+/// bytes and memo data bytes. High-level memo objects should be convertible
+/// to MemoPayload. Deserialization, across all high-level memo types, is
+/// done in mc-memo crate.
+///
+/// Note that a memo payload may be invalid / uninterpretable, or refer to new
+/// memo types that have been introduced at a later date.
+#[derive(Clone, Deserialize, Default, Eq, Ord, PartialEq, PartialOrd, Serialize, Debug)]
+pub struct MemoPayload(GenericArray<u8, U34>);
```

Comment:
Why does this need to be a generic array and not `[u8; 34]`?

Reply:
for arrays over 32, you can't derive all these traits unless you use generic array, that's one of the main annoying things about the built-in arrays right now, which hopefully they will fix soon

Reply:
Doesn';t https://blog.rust-lang.org/2020/10/08/Rust-1.47.html#traits-on-larger-arrays address this?

Reply:
Maybe not for serde... :/

Reply:
@eranrund for now i'm just trying to keep this the same or similar to `EncryptedFogHint`, maybe we can make a ticket to switch all those to `[u8; n]` in a future PR if it doesn't need to happen right now